p2panda-rs/src/graph/traits.rs

```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+/// Trait to be defined on a reducer which can be passed into `Graph` in order to perform an
+/// operation on all visited nodes.
+///
+/// It has one method `combine` which takes a generic value and mutable self.
+pub trait Reducer<V> {
+    /// Takes a generic value and presumably combines it with some contained state.
+    fn combine(&mut self, value: &V);
```

Comment:
I'd like to explore if this should be able to return exceptions, maybe we make an issue for this?

Reply:
Ah yeh, could actually just implement that in this PR. Sure there are cases where an error is never expected, but that's no problem. The current pattern when reducing a document and checking the reducer for errors at the end is a bit weird.

Reply:
How's that look now, with `combine` returning an error?

Reply:
Looking super :+1: